sympy/solvers/solvers.py

```diff
@@ -863,64 +876,124 @@ def _sympified_list(w):
+            symbols = symbols[0]
+            if isinstance(symbols, GeneratorType):
+                symbols = list(symbols)
+        # Has the user given an unordered collection like a set?
+        ordered_symbols = is_sequence(symbols)
+        symbols = [sympify(sym) for sym in symbols]
+
+    # Remove symbols the user is not interested in. This removes symbols even
+    # if they were explicitly specified.
 
-    # remove symbols the user is not interested in
-    exclude = flags.pop('exclude', set())
     if exclude:
-        if isinstance(exclude, Expr):
+        if not iterable(exclude):
             exclude = [exclude]
         exclude = set().union(*[e.free_symbols for e in sympify(exclude)])
-    symbols = [s for s in symbols if s not in exclude]
+        symbols = [s for s in symbols if s not in exclude]
 
+    # alternate exits
 
-    # preprocess equation(s)
-    ###########################################################################
-    for i, fi in enumerate(f):
-        if isinstance(fi, (Equality, Unequality)):
-            if 'ImmutableDenseMatrix' in [type(a).__name__ for a in fi.args]:
-                fi = fi.lhs - fi.rhs
-            else:
-                L, R = fi.args
-                if isinstance(R, BooleanAtom):
-                    L, R = R, L
-                if isinstance(L, BooleanAtom):
-                    if isinstance(fi, Unequality):
-                        L = ~L
-                    if R.is_Relational:
-                        fi = ~R if L is S.false else R
-                    elif R.is_Symbol:
-                        return L
-                    elif R.is_Boolean and (~R).is_Symbol:
-                        return ~L
-                    else:
-                        raise NotImplementedError(filldedent('''
-                            Unanticipated argument of Eq when other arg
-                            is True or False.
-                        '''))
+    if any(_is_eq_bool(fi) for fi in f):
+        for i, fi in enumerate(f):
+            if not isinstance(fi, (Equality, Unequality)):
+                continue
+            args = fi.args
+            if args[1] in (S.true, S.false):
+                args = args[1], args[0]
+            L, R = args
+            if L in (S.false, S.true):
+                if isinstance(fi, Unequality):
+                    L = ~L
+                if R.is_Relational:
+                    pass
+                elif R.is_Symbol:
+                    return L
+                elif R.is_Boolean and (~R).is_Symbol:
+                    return ~L
                 else:
-                    fi = fi.rewrite(Add, evaluate=False)
-            f[i] = fi
+                    raise NotImplementedError(filldedent('''
+                        Unanticipated argument of Eq when other arg
+                        is True or False.
+                    '''))
+
+    f_relational = [_invert_eq_bool(fi) for fi in f]
+
+    if any(_is_relational(fi) for fi in f_relational):
```

Comment:
Can we move everything inside this block to a separate function?

Reply:
Yeah, just did